lib/phoenix/pubsub/local.ex

```diff
@@ -129,39 +173,90 @@ defmodule Phoenix.PubSub.Local do
 
   @doc false
   # This is an expensive and private operation. DO NOT USE IT IN PROD.
-  def list(local_server) when is_atom(local_server) do
-    local_server
+  def list(pubsub_server, shard) when is_atom(pubsub_server) do
+    shard
+    |> local_for_shard(pubsub_server)
     |> :ets.select([{{:'$1', :_}, [], [:'$1']}])
     |> Enum.uniq
   end
 
   @doc false
-  # This is an expensive and private operation. DO NOT USE IT IN PROD.
-  def subscription(local_server, pid) when is_atom(local_server) do
-    local_server
-    |> :ets.select([{{:'$1', {pid, :_}}, [], [:'$1']}])
-    |> Enum.uniq
+  def subscription(pubsub_server, pool_size, pid) when is_atom(pubsub_server) do
+    pid
+    |> pid_to_shard(pool_size)
+    |> local_for_shard(pubsub_server)
+    |> GenServer.call({:subscription, pid})
+  end
+
+  @doc false
+  def local_name(pubsub_server, shard) do
+    Module.concat(["#{pubsub_server}.Local#{shard}"])
+  end
+
+  @doc false
+  def gc_name(pubsub_server, shard) do
+    Module.concat(["#{pubsub_server}.GC#{shard}"])
   end
 
   def init({local, gc}) do
     ^local = :ets.new(local, [:duplicate_bag, :named_table, :public,
                               read_concurrency: true, write_concurrency: true])
+    ^gc = :ets.new(gc, [:duplicate_bag, :named_table, :public,
+                        read_concurrency: true, write_concurrency: true])
+
     Process.flag(:trap_exit, true)
-    {:ok, gc}
+    {:ok, %{topics: local, pids: gc, gc_server: gc}}
   end
 
-  def handle_call({:subscribe, pid, link}, _from, state) do
-    if link, do: Process.link(pid)
+  def handle_call({:subscribe, pid, _topic, opts}, _from, state) do
+    if opts[:link], do: Process.link(pid)
     Process.monitor(pid)
-    {:reply, :ok, state}
+    {:reply, {:ok, {state.topics, state.pids}}, state}
+  end
+
+  def handle_call({:subscription, pid}, _from, state) do
+    try do
+      {:reply, :ets.lookup_element(state.pids, pid, 2), state}
```

Comment:
Ah, if you see my earlier commits, the one that introduced ETS, I have worked around this by sending a fake request to the server. Sometimes a subscribe, just to ensure it has caught up.